### PR TITLE
Fix S3937 FP and other improvements on number pattern checks

### LINKED ISSUE #3165: FP S3937: Hexadecimal patterns are not handled properly
### Description
For hexadecimal defined numbers (int and long), groups of 3 are okay, but 2 and 4 are not. I think that is not what should be the case.

### Expected behavior
``` C#
public static class S3937
{
    public static readonly long Hexa2 = 0xFF_FF_FF_FF; // compliant
    public static readonly long Hexa3 = 0x121_1F3_456; // non-compliant, hexadecimals should have groups of 2 or 4 
    public static readonly long Hexa4 = 0xF_FFFF_FFFF; // compliant
}
```
### Already okay
The binary defined work like I expected.
``` C#
public static class S3937
{
    public static readonly long Bin2 = 0b1_10_11_10_11; // compliant
    public static readonly long Bin3 = 0b0_111_111_111; // non-compliant, binaries should have groups of 2 or 4 
    public static readonly long Bin4 = 0b111_1101_1011; // compliant
}
```
* SonarC# Version 8.3

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #3165

Take binary and hexadecimal prefixes into account.

All groups should have the same size with these exceptions:
* first group before the dot is allowed to be shorter.
* last group after the dot is allowed to be shorter.
* It is allowed to only have groups before the dot.
* the first group after the dot (if multiple) should be one shorter than the group size

Specially the last one was missing.
``` C#
var amout = 1_123_123.23_123_12m; 
```

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 1]
```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            {
+                length -= 2;
+            }
+            // single suffixes
+            else if ("LDFUMldfum".IndexOf(numericToken[numericToken.Length - 1]) != NotFound)
+            {
+                length--;
+            }
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: This part does not solve the issue that there will be FP for hexadecimal numbers finishing with `D` or `F`.
Ex:
```
var a = 0xFF_FF_1F; // FP
var a = 0xFF_FF_1D; // FP
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 2]
```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
...
+
+            var decimals = splitted[1].Split(Underscore).Select(s => s.Length).ToArray();
+
+            if (decimals.Length == 1)
+            {
+                return false;
+            }
+
+            // The first decimal group should be one smaller.
+            decimals[0]++;
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Why this change in the rule behaviour? I'm not saying this is necessarily wrong, but if we change that, it first should be updated in the rule description

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 3]
```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            if (NumericTypeSuffix.Contains(char.ToUpperInvariant(numberLiteral[numberLiteral.Length - 1])))
+            // all should have the same size, except the last.
+            if (decimals.Take(decimals.Length - 1).Any(g => g != size))
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 1);
+                return true;
             }
 
-            return numberLiteral;
+            return false;
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: this part can be simplified with:
```
return decimals.Last() > size || decimals.Take(decimals.Length - 1).Any(g => g != size);
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 4]
```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        public static bool HasIrregularPattern(string numericToken)
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: I feel this method is getting quite big. I would have kept a separate utility method to retrieve the actual content of the numeric token, to have that logic separated from the pattern check

Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Method visibility could be kept `internal` as we only want to access it from outside for testing purposes

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 1 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Well, given the discussion I was under the impression that we where not going for groups of 3 for decimals and groups of 2, 4, or 8 for hex and binary. I can live with both, but there were at least some in consistencies (and bugs), that had to be fixed. If there be found agreement on the allowed group sizes, its is easy to add.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 3 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Shorter !- simpler.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 2 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: I think is the only right implementation. Given this changes, some descriptions have to be updated anyway.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 4 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: If you want to do something with the decimal/hexa/binary info, that is becoming less optimal.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 1 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Yes, I still need to discuss further with another colleague regarding the rule but for now let's assume we don't want a change in the behaviour, as this could be done in a separate PR if needed.

The false positive in the rule I'm talking about is not directly linked to the number of elements of such groups: it is related to the fact that hexedecimal numbers finishing with a `F` or `D` do not behave as expected.

Current behaviour is the following:
```
var a = 0xFF_FF_F1; // Ok - no issue raised, groups of 2
var b = 0xFFF_FFF_FF1; // Ok - no issue raised, groups of 3
var c = 0xFFFF_FFFF_FFF1; // Ok - no issue raised, groups of 4

var d = 0xFF_FF_FF; // KO - issue raised but should not - last group is thought to be of size 1 as the last 'F' of the group is considered as float number suffix - FP
var e = 0xFFF_FFF_FFD; // KO - issue raised but should not - last group is thought to be of size 2 as the last 'D' of the group is considered as double number suffix - FP
var f = 0xFFFF_FFFF_FFFF; // KO - issue raised but should not - FP

var g = 0xFF_FF_FFF; // No issue raised, but it should - grouping is not consistant
var h = 0xFFF_FFF_FFFD; // No issue raised, but it should - grouping is not consistant
var i = 0xFFFF_FFFF_FFFFF; // No issue raised, but it should - grouping is not consistant
```
This is the most important problem with the rule at the moment

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 2 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Could you direct me to some documentation regarding this ? I cannot find anything on the grouping of the fractional part to justify this change

Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: @christophe-zurn-sonarsource No, I can't. :(

I'm pretty sure I've been educated that way, but I can't find any source that says I'm right.

it is probably only true if the group size is 3 anyway. What I remember is that the first decimal are 1/10, then 1/100, so it makes sense to add a _ between 1/100 and 1/1000, just like between 1000 and 100.

But I agree, that without proof, we should not go for it.

The only thing I found (that also slightly disagrees) is this:
https://math.stackexchange.com/questions/182775/convention-of-digit-grouping-after-decimal-point/182776#182776

What suggests that the group size of your decimal part can be different (although it still should be consequent).

**Comment by andrei-epure-sonarsource** [comment]
tests are failing on this one (see [here](https://sonarsource.visualstudio.com/DotNetTeam%20Project/_build/results?buildId=16459&view=logs&j=22b521bd-d457-5294-620e-7d91dbfb4c56&t=74d1794f-b183-5917-cd98-3eaf49a73ed6&l=1981)), you should rebase on master

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 2 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: I've changed the behaviour. Also I added that the group size should be at least, two, as I think that makes sense. ;)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 1 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: @christophe-zurn-sonarsource 

The false positives have been fixed, and a minium group size of 2 is also taken into account.

**Comment by andrei-epure-sonarsource** [comment]
as long as the build is broken, we cannot move forward, @Corniel

**Comment by Corniel** [comment]
@ andrei-epure-sonarsource Any clue what is going wrong? maybe a new rebase?

**Comment by andrei-epure-sonarsource** [comment]
not sure @Corniel , I think one of us mentioned a failing test similar to what's failing on this PR.

we're working on a different project this sprint so most likely we'll investigate it a bit later on

enjoy the long weekend!

**Comment by Corniel** [comment]
@andrei-epure-sonarsource Same to you guys!

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs [hunk 5]
```diff
@@ -46,5 +46,37 @@ public void NumberPatternShouldBeRegular_FromCSharp7()
...
+
+        [DataTestMethod]
+        [TestCategory("Rule")]
+        [DataRow(".123_123_123_1", "No group before the dot")]
+        [DataRow("123", "No group character")]
+        [DataRow("1_123_123UL", "With UL suffix")]
+        [DataRow("1_123_123L", "With L suffix")]
+        [DataRow("1.1.1", "Two dots")]
+        [DataRow("0b1010_1010", "With binary prefix")]
+        [DataRow("0xFF_FF_12", "With hexadecimal prefix")]
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs:71]: `0xFFFF_FFFF_FFFF` is considered irregular because you strip the `F` suffix

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs [hunk 6]
```diff
@@ -46,5 +46,37 @@ public void NumberPatternShouldBeRegular_FromCSharp7()
...
         }
+
+        [DataTestMethod]
+        [TestCategory("Rule")]
+        [DataRow("1_1_1", "Group size of 1")]
+        [DataRow("123_12", "First group is bigger then the second")]
+        [DataRow("1234_123_123", "First group is bigger then the others")]
+        [DataRow("123_123_12_123", "Another group has a different size")]
+        [DataRow("1_123.123_1234", "Last decimal group is bigger")]
+        [DataRow(".123_123_1234", "No group before the dot")]
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs:57]: `0xFF_FF_FFF` is considered regular because you strip the `F` suffix

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 7]
```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
...
+
+            var length = numericToken.Length - start;
+
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            {
+                length -= 2;
+            }
+            // single suffixes
+            else if ("LDFUMldfum".IndexOf(numericToken[numericToken.Length - 1]) != NotFound)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:134]: you should only strip the `F` and `D` suffix for non-binary, non-hexa numbers as they are hexa "digits"

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 8]
```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
...
+        private static string StripNumericPreAndSuffix(string numericToken)
+        {
+            // hexadecimal en binary prefixes
+            var start = numericToken.StartsWith("0x", StringComparison.InvariantCultureIgnoreCase)
+                || numericToken.StartsWith("0b", StringComparison.InvariantCultureIgnoreCase) ? 2 : 0;
+
+            var length = numericToken.Length - start;
+
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:129]: it can also end with "LU" ([docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types))

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs [hunk 9]
```diff
@@ -46,5 +46,37 @@ public void NumberPatternShouldBeRegular_FromCSharp7()
...
+        public void HasIrregularPattern(string numericToken, string message)
+        {
+            Assert.IsTrue(NumberPatternShouldBeRegular.HasIrregularPattern(numericToken), message);
+        }
+
+        [DataTestMethod]
+        [TestCategory("Rule")]
+        [DataRow(".123_123_123_1", "No group before the dot")]
+        [DataRow("123", "No group character")]
+        [DataRow("1_123_123UL", "With UL suffix")]
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs:67]: also add with "LU" suffix as it's legal, also to test case insensitivity
```
1_123_123LU
1_123_123lU
1_123_123Lu
1_1234_1234lu
1_12_13ul
```
and with [floating point suffixes](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types) (the below numbers are from the MS docs)

```
1.0
3D
0.42e2
3_000.5F
134.45E-2f
3_000.5m
1.5E6m
2.1m
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 10]
```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
...
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        internal static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            var splitted = StripNumericPreAndSuffix(numericToken).Split(Dot);
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:71]: please use `split` (the participle and past tense forms are the same as the present form)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 11]
```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            }
+
+            if (splitted.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 2);
+                return false;
             }
 
-            if (NumericTypeSuffix.Contains(char.ToUpperInvariant(numberLiteral[numberLiteral.Length - 1])))
+            var decimals = splitted[1].Split(Underscore).Select(s => s.Length).ToArray();
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:105]: do you mean `digits`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 12]
```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            if (decimals.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 1);
+                return false;
             }
 
-            return numberLiteral;
+            size = size == NotFound ? decimals[0] : size;
+
+            // the last should not be bigger
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:114]: by "last" do you mean first? you mean that `0xF_FF_FF` is ok even if the first group has a different number of digits, right?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 7 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:134]: also, how do we treat 'e' - what's after 'e' should be like a suffix, right?

```
1.5E6m
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs [hunk 13]
```diff
@@ -46,7 +46,7 @@ public void NonCompliantCases()
             int million = 1_000_00_000;  // Noncompliant {{Review this number; its irregular pattern indicates an error.}}
 //                        ^^^^^^^^^^^^
 
-            var x = 1_234.56_78; // Noncompliant
```
[the remark this answers is not in the document]
Comment (Corniel) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:49]: 1_234.56_78 According to the sources I found, there is no reason why the decimals should have the same group size, as the-non-decimals. So it should be compliant.

[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:49]: ok, sounds good to me, so please leave the test as `// Compliant` 😄 

the main point here is _please don't remove an existing test case_

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs:57]: You're right. Fixed!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 7 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:134]: I Fully agree. Will add that too!

Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:139]: Hmm, seconds thoughts:
1.234_567E2 Is obviously valid
but:
1.234_56E2
Is that wrong? should it be 0.123.456E3

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:49]: @andrei-epure-sonarsource IK get your point, but the thing is: It was compliant now, and therefore also on a different place. I changed that test to become non-compliant and left that at the same spot. So what do you suggest?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 12 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:115]: yep

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:49]: my apologies, I missed the `double groupsizes = 1_234.56_78;`

**Comment by andrei-epure-sonarsource** [comment]
there's a [failed UT](https://sonarsource.visualstudio.com/DotNetTeam%20Project/_build/results?buildId=18110&view=logs&j=23c99647-d7e1-5b83-b5fb-f7e97f8a9ce7&t=3ae69f40-e780-58ec-7b2f-ab41fb46316e&l=49) 

>   X NumberPatternShouldBeRegular_FromCSharp7 [29ms]
  Error Message:
   Test method SonarAnalyzer.UnitTest.Rules.NumberPatternShouldBeRegularTest.NumberPatternShouldBeRegular_FromCSharp7 threw exception: 
SonarAnalyzer.UnitTest.TestFramework.UnexpectedDiagnosticException: Unexpected primary issue on line 15, span (14,32)-(14,43) with message 'Review this number; its irregular pattern indicates an error.'.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 14]
```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
...
+        {
+            var split = StripNumericPreAndSuffix(numericToken).Split(Dot);
+
+            // ignore multiple dots.
+            if (split.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = split[0].Split(Underscore).Select(s => s.Length).ToArray();
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:80]: I suggest to rename to `groupLengths`
- this isn't "groups of digits", but the size of each groups
- it doesn't refer to all groups, but only to the ones on the left of the 'dot' (if the 'dot' is present) - *please add a comment above to mention something like* `// we are interested only in what is on the left side of the 'Dot', if present`

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 15]
```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                size = groups[1];
+
+                // first should not be bigger.
+                if (groups[0] > size)
+                {
+                    return true;
+                }
+
+                // all should have the same size, except the first, and the size should at least be 2.
+                if (size < 2 || groups.Skip(1).Any(g => g != size))
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:95]: instead of `g` maybe use `l` (for `lenght`)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 16]
```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                    return true;
+                }
+            }
+
+            if (split.Length == 1)
+            {
+                return false;
+            }
+
+            var decimals = split[1].Split(Underscore).Select(s => s.Length).ToArray();
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:106]: a [decimal](https://dictionary.cambridge.org/dictionary/english/decimal) is 
> relating to or expressed in a system of counting based on the number ten

please rename to [digit ](https://en.wiktionary.org/wiki/digit#Noun), as it can be in hexa as well

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 17]
```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                }
+
+                // all should have the same size, except the first, and the size should at least be 2.
+                if (size < 2 || groups.Skip(1).Any(g => g != size))
+                {
+                    return true;
+                }
+            }
+
+            if (split.Length == 1)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:101]: please do this in the same `if` with `split.Length > 2` , immediately after the split

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 18]
```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            {
+                return false;
+            }
+
+            var decimals = split[1].Split(Underscore).Select(s => s.Length).ToArray();
+
+            if (decimals.Length == 1)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:108]: I'me really confused here - haven't you done this check already, if the group has 1 digit?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NumberPatternShouldBeRegularTest.cs:80]: can you also add `0xFFFF_FFFF_FFFF` please as a regular format?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 16 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:106]: also, as above - this if `digitGroupSizes` rather than the digits themselves

Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:106]: and again this is a bit weird - why are you splitting based on Undescore here and at line 80 as well?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 12 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: Changed this code to make the intent clear. Thanks for the indebt code-reviews. It has been a long time since my code has been checked that well!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 17 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:88]: Done.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 19]
```diff
@@ -150,5 +129,24 @@ private static string StripNumericPreAndSuffix(string numericToken)
...
+        {
+            if (groupLengths.Length < 2)
+            {
+                return false;
+            }
+
+            // the second group is assumed to be leading.
+            var groupLength = groupLengths[1];
+
+            // first should not be bigger, and the size per group should be at least 2.
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:143]: this comment is a bit harder to read because you say "first should not be bigger" but that's the second expression in the `OR` expression

```suggestion
            // we consider groups of 1 digit irregular
            // the first group should not be bigger than the rest
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs [hunk 20]
```diff
@@ -150,5 +129,24 @@ private static string StripNumericPreAndSuffix(string numericToken)
...
         }
+
+        private static bool HasIrregularGroupLengths(int[] groupLengths)
+        {
+            if (groupLengths.Length < 2)
+            {
+                return false;
+            }
+
+            // the second group is assumed to be leading.
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs:140]: I don't understand what "leading" means?

```suggestion
            // the first group is allowed to contain less digits that the other ones, so take the "regular" length from the second group
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs [hunk 21]
```diff
@@ -47,7 +47,7 @@ public void NonCompliantCases()
             int million = 1_000_00_000;  // Noncompliant {{Review this number; its irregular pattern indicates an error.}}
 //                        ^^^^^^^^^^^^
 
-            var x = 1_234.56_78_1; // Noncompliant
+            var x = 1_234.56_78_123; // Noncompliant
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:51]: do we have any tests with `____`? it would be nice to have something like `0xFF___FF___FF` - I believe it's Noncompliant due to line 144 in `NumberPatternShouldBeRegular.cs`

and I think it's fine to be noncompliant (even if the pattern is regular, but using multiple underscores should usually be a mistake)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs (hunk 12 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs]: I'm glad you appreciate the comments and thanks for the patience :)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs (hunk 21 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:51]: and also with `0xFF________FF___FF`

this program runs:
```
var x = 0xFF________FF___FF;
Console.WriteLine(x);
```

Comment (Corniel) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/NumberPatternShouldBeRegular.cs:51]: Added the tests, and took the comment suggestions.
